src/Atmos/Model/AtmosModel.jl

```diff
@@ -896,4 +898,171 @@ function numerical_flux_first_order!(
+    NumericalFluxFirstOrder()
+        ::HLLCNumericalFlux,
+        balance_law::AtmosModel,
+        fluxᵀn,
+        normal_vector,
+        state_prognostic⁻,
+        state_auxiliary⁻,
+        state_prognostic⁺,
+        state_auxiliary⁺,
+        t,
+        direction,
+    )
+
+An implementation of the numerical flux based on the HLLC method for
+the AtmosModel. For more information on this particular implementation,
+see Chapter 10.4 in the provided reference below.
+
+## References
+    @book{toro2013riemann,
+        title={Riemann solvers and numerical methods for fluid dynamics: a practical introduction},
+        author={Toro, Eleuterio F},
+        year={2013},
+        publisher={Springer Science & Business Media}
+    }
+"""
+function numerical_flux_first_order!(
+    ::HLLCNumericalFlux,
+    balance_law::AtmosModel,
+    fluxᵀn::Vars{S},
+    normal_vector::SVector,
+    state_prognostic⁻::Vars{S},
+    state_auxiliary⁻::Vars{A},
+    state_prognostic⁺::Vars{S},
+    state_auxiliary⁺::Vars{A},
+    t,
+    direction,
+) where {S, A}
+    FT = eltype(fluxᵀn)
+    num_state_prognostic = number_states(balance_law, Prognostic())
+    param_set = balance_law.param_set
+
+    # Extract the first-order fluxes from the AtmosModel (underlying BalanceLaw)
+    # and compute normals on the positive + and negative - sides of the
+    # interior facets
+    flux⁻ = similar(parent(fluxᵀn), Size(3, num_state_prognostic))
+    fill!(flux⁻, -zero(FT))
+    flux_first_order!(
+        balance_law,
+        Grad{S}(flux⁻),
+        state_prognostic⁻,
+        state_auxiliary⁻,
+        t,
+        direction,
+    )
+    fluxᵀn⁻ = flux⁻' * normal_vector
+
+    flux⁺ = similar(flux⁻)
+    fill!(flux⁺, -zero(FT))
+    flux_first_order!(
+        balance_law,
+        Grad{S}(flux⁺),
+        state_prognostic⁺,
+        state_auxiliary⁺,
+        t,
+        direction,
+    )
+    fluxᵀn⁺ = flux⁺' * normal_vector
+
+    # Extract relevant fields and thermodynamic variables defined on
+    # the positive + and negative - sides of the interior facets
+    ρ⁻ = state_prognostic⁻.ρ
+    ρu⁻ = state_prognostic⁻.ρu
+    ρe⁻ = state_prognostic⁻.ρe
+    ts⁻ = thermo_state(
+        balance_law,
+        balance_law.moisture,
+        state_prognostic⁻,
+        state_auxiliary⁻,
+    )
+
+    u⁻ = ρu⁻ / ρ⁻
+    c⁻ = soundspeed_air(ts⁻)
+
+    uᵀn⁻ = u⁻' * normal_vector
+    e⁻ = ρe⁻ / ρ⁻
+    p⁻ = pressure(
+        balance_law,
+        balance_law.moisture,
+        state_prognostic⁻,
+        state_auxiliary⁻,
+    )
+
+    ρ⁺ = state_prognostic⁺.ρ
+    ρu⁺ = state_prognostic⁺.ρu
+    ρe⁺ = state_prognostic⁺.ρe
+    ts⁺ = thermo_state(
+        balance_law,
+        balance_law.moisture,
+        state_prognostic⁺,
+        state_auxiliary⁺,
+    )
+
+    u⁺ = ρu⁺ / ρ⁺
+    uᵀn⁺ = u⁺' * normal_vector
+    e⁺ = ρe⁺ / ρ⁺
+    p⁺ = pressure(
+        balance_law,
+        balance_law.moisture,
+        state_prognostic⁺,
+        state_auxiliary⁺,
+    )
```

Comment:
```suggestion
    p⁺ = air_pressure(ts⁺)
```

Reply:
I'm now thinking that we should remove `pressure`, since this relies on `recover_thermo_state` and will have the same issue as above. Using `ts⁺ = new_thermo_state(...)` will fix these issues.

Reply:
Done